### PR TITLE
nixos/dropbox: new module

### PR DESCRIPTION
**Pull request by romildo**
###### Motivation for this change

Add a module for the dropbox service.

See also https://nixos.wiki/wiki/Dropbox

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/dropbox.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      Whether to enable the dropbox service globally.
+    '';
+
+    install = mkEnableOption ''
+      Whether to install a user service for the dropbox daemon. Once
+      the service is started, the Dropbox desktop app is available.
+
+      The service must be manually started for each user with
+      "systemctl --user start dropbox" or globally through
+      <varname>services.emacs.enable</varname>.
```
Comment (schmittlauch) [inline nixos/modules/services/networking/dropbox.nix:25]: I guess this is a copy-paste leftover, or since when does Dropbox use emacs as a backend? :D

**Comment by peterhoeg** [comment]
Since we have an open source alternative, `maestral`, we should make the package used configurable to support both.

File: nixos/modules/services/networking/dropbox.nix (hunk 1 above)
Comment (romildo) [reply nixos/modules/services/networking/dropbox.nix]: Fixed.

**Comment by romildo** [comment]
> Since we have an open source alternative, `maestral`, we should make the package used configurable to support both.

The systemd units would be different for `dropbox` and for `maestral`. Maybe a solution is to let the package install the unit, and the module would use it.

**Comment by peterhoeg** [comment]
>   The systemd units would be different for dropbox and for maestral.

Sure.

> Maybe a solution is to let the package install the unit, and the module
> would use it.

That's definitely an option, but you would still have to augment it if you want to activate it globally so you still need to handle the different units.

File: nixos/modules/services/networking/dropbox.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  config = mkIf (cfg.enable || cfg.install) {
+
+    environment.systemPackages = with pkgs; [ dropbox-cli ];
+
+    systemd.user.services.dropbox = {
+      description = "Dropbox daemon";
+      after = [ "xembedsniproxy.service" ];
+      wants = [ "xembedsniproxy.service" ];
+      environment = {
+        QT_PLUGIN_PATH = "/run/current-system/sw/" + pkgs.qt5.qtbase.qtPluginPrefix;
```
Comment (jtojnar) [inline nixos/modules/services/networking/dropbox.nix:48]: Could you add a comment why this is needed? Should not the app be wrapped with `wrapQtAppsHook`?

File: nixos/modules/services/networking/dropbox.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  ###### implementation
+
+  config = mkIf (cfg.enable || cfg.install) {
+
+    environment.systemPackages = with pkgs; [ dropbox-cli ];
+
+    systemd.user.services.dropbox = {
+      description = "Dropbox daemon";
+      after = [ "xembedsniproxy.service" ];
```
Comment (jtojnar) [inline nixos/modules/services/networking/dropbox.nix:45]: Huh?

File: nixos/modules/services/networking/dropbox.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      wants = [ "xembedsniproxy.service" ];
+      environment = {
+        QT_PLUGIN_PATH = "/run/current-system/sw/" + pkgs.qt5.qtbase.qtPluginPrefix;
+        QML2_IMPORT_PATH = "/run/current-system/sw/" + pkgs.qt5.qtbase.qtQmlPrefix;
+      };
+      serviceConfig = {
```
Comment (jtojnar) [inline nixos/modules/services/networking/dropbox.nix:51]: Upstream service file would be indeed nicer. How do other distros handle this?

**Comment by peterhoeg** [comment]
>   Upstream service file would be indeed nicer. How do other distros handle
>   this?

Neither dropbox nor maestral upstream carry systemd units.

In maestral's case it's been on my list for a long time but until that happens, we'll need to handle it here.

File: nixos/modules/services/networking/dropbox.nix (hunk 2 above)
Comment (Mic92) [reply nixos/modules/services/networking/dropbox.nix:48]: I think dropbox downloads its own binaries at runtime.

File: nixos/modules/services/networking/dropbox.nix (hunk 3 above)
Comment (romildo) [reply nixos/modules/services/networking/dropbox.nix:45]: @peterhoeg can you comment on the dependency on `xembedsniproxy.service`? What is its purpose?

Comment (peterhoeg) [reply nixos/modules/services/networking/dropbox.nix:45]: It used to be the only way to have dropbox show a trayicon - I have no idea if that is still the case as I'm not using dropbox any more.

Comment (jtojnar) [reply nixos/modules/services/networking/dropbox.nix:45]: Looks like some kind of plasma thing: https://github.com/davidedmundson/xembed-sni-proxy

**Comment by SRGOM** [comment]
@peterhoeg Peter, might I take your time/attention to get #77734 (re) reviewed too? Along the same lines (and while I don't know the policy regarding what belongs in Nixos and what doesn't, and the policy changes related to that, I feel that dropbox and OneDrive would be under the same policy...)

**Pull request closed without merging**